src/array.jl

```diff
@@ -0,0 +1,163 @@
+end
+
+@compat function CLArray{T}(buf::Buffer{T}, sz::Tuple{Vararg{Int}})
+    ctx = context(buf)
+    queue = CmdQueue(ctx)
+    CLArray(context(buf), queue, buf, sz)
+end
+
+Base.copy(A::CLArray; ctx=A.ctx, queue=A.queue, buffer=A.buffer, size=A.size) =
+    CLArray(ctx, queue, buffer, size)
+function Base.deepcopy{T,N}(A::CLArray{T,N})
+    new_buf = Buffer(T, A.ctx, prod(A.size))
+    copy!(A.queue, new_buf, A.buffer)
+    return CLArray(A.ctx, A.queue, new_buf, A.size)
+end
+
+
+"""
+Create in device memory array of type `t` and size `dims` filled by value `x`.
+"""
+function Base.fill{T}(::Type{T}, q::CmdQueue, x::T, dims...)
+    ctx = info(q, :context)
+    v = opencl_version(ctx)
+    if v.major == 1 && v.minor >= 2
+        buf = Buffer(T, ctx, prod(dims))
+        fill!(q, buf, x)
+    else
+        buf = Buffer(T, ctx, (:rw, :copy), prod(dims), hostbuf=fill(x, dims))
+    end
+    return CLArray(buf, dims)
+end
+
+Base.zeros{T}(::Type{T}, q::CmdQueue, dims...) = fill(T, q, T(0), dims...)
+Base.zeros(q::CmdQueue, dims...) = fill(Float64, q, Float64(0), dims...)
+Base.ones{T}(::Type{T}, q::CmdQueue, dims...) = fill(T, q, T(1), dims...)
+Base.ones(q::CmdQueue, dims...) = fill(Float64, q, Float64(1), dims...)
+
+
+##  core functions
+
+buffer(A::CLArray) = A.buffer
+bufptr(A::CLArray) = A.buffer.id
+context(A::CLArray) = context(A.buffer)
+queue(A::CLArray) = context(A.queue)
+Base.size(A::CLArray) = A.size
+Base.ndims(A::CLArray) = length(size(A))
+Base.length(A::CLArray) = prod(size(A))
+Base.(:(==))(A:: CLArray, B:: CLArray) =
+    buffer(A) == buffer(B) && size(A) == size(B)
+Base.reshape(A::CLArray, dims...) = begin
+    @assert prod(dims) == prod(size(A))
+    return copy(A, size=dims)
+end
+
+##  show
+
+Base.show{T,N}(io::IO, A::CLArray{T,N}) =
+    print(io, "CLArray{$T,$N}($(buffer(A)),$(size(A)))")
+
+##  to_host
+
+function to_host{T,N}(A::CLArray{T,N}; queue=A.queue)
+    hA = Array(T, size(A))
+    copy!(queue, hA, buffer(A))
+    return hA
+end
+
+## kernels
+
+"""
+Format string using dict-like variables, replacing all accurancies of
+`%(key)` with `value`.
+
+Example:
+    s = "Hello, %(name)"
+    format(s, name="Tom")  ==> "Hello, Tom"
+"""
+function format(s::AbstractString; vars...)
+    for (k, v) in vars
+        s = replace(s, "%($k)", v)
+    end
+    s
+end
+
+const PROGRAM_TRANSPOSE = readall(Pkg.dir("OpenCL/src/kernels/transpose.cl"))
+
+function build_kernel(ctx::Context, program::AbstractString,
+                      kernel_name::AbstractString; vars...)
+    src = format(PROGRAM_TRANSPOSE; vars...)
+    p = Program(ctx, source=src)
+    build!(p)
+    return Kernel(p, kernel_name)
+end
+
+## other array operations
+
+"""Transpose CLMatrix A, write result to a preallicated CLMatrix B"""
+function Base.transpose!(B::CLMatrix{Float32}, A::CLMatrix{Float32};
+                         queue=A.queue, block_size=32)
+    ctx = context(A)
+    kernel = build_kernel(ctx, PROGRAM_TRANSPOSE, "transpose",
```

Comment:
We should have a simple cache here that saves the compiled kernel for a given block size.